pkg/datapath/connector/config.go

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package connector
+
+import (
+	"fmt"
+	"log/slog"
+	"math"
+
+	"github.com/cilium/hive/cell"
+
+	datapathOption "github.com/cilium/cilium/pkg/datapath/option"
+	"github.com/cilium/cilium/pkg/datapath/tunnel"
+	"github.com/cilium/cilium/pkg/datapath/types"
+	"github.com/cilium/cilium/pkg/option"
+	wgTypes "github.com/cilium/cilium/pkg/wireguard/types"
+)
+
+// Connector configuration. As per BIGTCP, the values here will not be calculated
+// until the Hive has started. This is necessary to allow other dependencies to
+// setup their interfaces etc.
+type ConnectorConfig struct {
+	UserConfig types.ConnectorUserConfig
+
+	// podDeviceHeadroom tracks the desired headroom buffer margin for the
+	// network device pair facing a workload.
+	podDeviceHeadroom uint16
+
+	// podDeviceTailroom tracks the desired tailroom buffer margin for the
+	// network device pairs facing a workload.
+	podDeviceTailroom uint16
+}
+
+func (cc *ConnectorConfig) GetPodDeviceHeadroom() uint16 {
+	return cc.podDeviceHeadroom
+}
+
+func (cc *ConnectorConfig) GetPodDeviceTailroom() uint16 {
+	return cc.podDeviceTailroom
+}
+
+type connectorParams struct {
+	cell.In
+
+	Lifecycle    cell.Lifecycle
+	Log          *slog.Logger
+	Orchestrator types.Orchestrator
+	UserConfig   types.ConnectorUserConfig
+	DaemonConfig *option.DaemonConfig
+	WgAgent      wgTypes.WireguardAgent
+	TunnelConfig tunnel.Config
+}
+
+// newConnectorConfig initialises a new ConnectorConfig object with default
+// parameters. If EnableTunedBufferMargins is enabled, we register a callback
+// with the Hive to detect necessary tuning parameters.
+func newConnectorConfig(p connectorParams) (*ConnectorConfig, error) {
+	cc := &ConnectorConfig{
+		UserConfig:        p.UserConfig,
+		podDeviceHeadroom: 0,
+		podDeviceTailroom: 0,
+	}
+
+	if cc.UserConfig.EnableTunedBufferMargins {
+		// At present, only netkit supports tuned buffer margins.
+		switch p.DaemonConfig.DatapathMode {
```

Comment:
Is there no way today to adjust `needed_headroom/needed_tailroom` on the fly for a veth pair? Skimming through [this series](https://lore.kernel.org/netdev/cover.1456412101.git.pabeni@redhat.com/) and looking at the kernel code it seems like `ndo_set_rx_headroom` is only ever invoked from OVS/bridge interfaces, but I just wanted to confirm. Does this mean that today `veth` when paired with WireGuard or a tunnel interface will require extra allocations+memcpys under the hood to expand the headroom?

Also, I'm curious if you have any numbers to compare before and after? How much overhead does this optimization remove?

Reply:
> I guess you could, but that feels very hacky and the complexity probably not worth it since we have it solved on netkit already.

I was thinking more in terms of environments running older kernels where netkit isn't an option. Anyway, just a thought, let's not try adding this now or anything.